pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java

```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
-        } else {
-          nextValueId = StarTreeIndexNode.all();
-        }
-        SearchEntry newEntry = new SearchEntry();
-        newEntry.starTreeIndexnode = current.getChildren().get(nextValueId);
-        if (newEntry.starTreeIndexnode != null) {
-          newEntry.remainingPredicates = newRemainingPredicates;
-          searchQueue.add(newEntry);
-        }
-      } else {
-        //if there is a group by
-        for (Map.Entry<Integer, StarTreeIndexNode> entry : current.getChildren().entrySet()) {
-          if (entry.getKey() != StarTreeIndexNode.all()) {
-            SearchEntry newEntry = new SearchEntry();
-            newEntry.starTreeIndexnode = entry.getValue();
-            newEntry.remainingPredicates = newRemainingPredicates;
-            searchQueue.add(newEntry);
+      // Find next set of nodes to search
+      String nextDimension =
+          dimensionIndexToNameMapping.inverse().get(current.getChildDimensionName());
+
+      HashSet<String> newRemainingPredicateColumns = new HashSet<>();
+      newRemainingPredicateColumns.addAll(remainingPredicateColumns);
+      HashSet<String> newRemainingGroupByColumns = new HashSet<>();
+      newRemainingGroupByColumns.addAll(remainingGroupByColumns);
+
+      addMatchingChildrenToQueue(searchQueue, current, nextDimension, newRemainingPredicateColumns,
+          newRemainingGroupByColumns);
+    }
+    return matchedEntries;
+  }
+
+  /**
+   * Helper method to add matching children into the search queue.
+   * - If predicate can be applied (i.e. equality predicate that is eligible), add the child
+   * satisfying
+   * the predicate into the queue.
+   * - If predicate cannot be applied (either inEligible or nonEquality), add all children to the
+   * queue.
+   * - If no predicate on the column, add the star-child to the queue
+   * @param searchQueue
+   * @param node
+   * @param column
+   * @param remainingPredicateColumns
+   * @param remainingGroupByColumns
+   */
+  private void addMatchingChildrenToQueue(Queue<SearchEntry> searchQueue, StarTreeIndexNode node,
+      String column, HashSet<String> remainingPredicateColumns,
+      HashSet<String> remainingGroupByColumns) {
+    Map<Integer, StarTreeIndexNode> children = node.getChildren();
+
+    if (equalityPredicateColumns.contains(column)) {
+      // Check if there is exact match filter on this column
+      int nextValueId;
+      PredicateEntry predicateEntry = predicatesMap.get(column);
+      nextValueId = predicateEntry.dictionaryId;
+      remainingPredicateColumns.remove(column);
+      remainingGroupByColumns.remove(column);
+      if (children.containsKey(nextValueId)) {
+        addNodeToSearchQueue(searchQueue, children.get(nextValueId), remainingPredicateColumns,
+            remainingGroupByColumns);
+      }
+    } else {
+      int nextValueId;
+      if (groupByColumns.contains(column) || predicatesMap.containsKey(column)
```

Comment:
Will it be more intuitive using remainingGroupByColumns and remainingPredicateColumns?